[WIP] Fix 3965

ibmjm opened the pull request:
Adds better output when failing to parse a datasource file for any reason, by including the filename of the datasource in question.

Fixes #3965

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [X] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

raymondfeng commented:
@ibmjm I have squashed all commits in your PR into one and fixed the commit message to comply with our conventional commit rules. See https://loopback.io/doc/en/lb4/submitting_a_pr.html

ibmjm commented:
> @ibmjm I have squashed all commits in your PR into one and fixed the commit message to comply with our conventional commit rules. See https://loopback.io/doc/en/lb4/submitting_a_pr.html

Thank you Raymond.  I did read the guide to submitting PRs prior to creating mine; which rule did I breach?

raymondfeng commented:
@ibmjm See https://loopback.io/doc/en/lb4/submitting_a_pr.html#1-commits-to-local-repository

raymondfeng commented:
https://loopback.io/doc/en/lb4/code-contrib-lb4.html#commit-message-guidelines

raymondfeng commented on packages/cli/lib/utils.js:
```diff
@@ -471,6 +471,7 @@ exports.getDataSourceConnectorName = function(datasourcesDir, dataSourceClass) {
     jsonFileContent = JSON.parse(fs.readFileSync(datasourceJSONFile, 'utf8'));
   } catch (err) {
     debug(`Error reading file ${datasourceJSONFile}: ${err.message}`);
+    err.message = `Cannot load ${datasourceJSONFile}: ${err.message}`;
```
Personally I would like to have a new Error that wraps the original one. It's easy to do so in Java, such as `new MyError(anotherError)`, but not in JavaScript :-(.

raymondfeng replied:
Something like https://javascript.info/custom-errors#wrapping-exceptions. Maybe we should introduce a helper in LB4.

ibmjm replied:
My first attempt to provide an error message that included the filename of the datasource where the failure originated, was indeed to throw a new Error.  However, @bajtos pointed out that we'd lose the stack trace by doing that -- hence the modification to the original Error's message.

raymondfeng replied:
I'm fine with your fix for now.

dreamdevil00 replied:
> Personally I would like to have a new Error that wraps the original one. It's easy to do so in Java, such as `new MyError(anotherError)`, but not in JavaScript :-(.

Have you considered using [verror](https://www.npmjs.com/package/verror)?
